### ISSUE TITLE
Optimize Output Error Model

### ISSUE THREAD

**Issue by arminwitte** [arminwitte/llsi#88]
Du triffst den Nagel auf den Kopf.
Offline-Identifikation (Batch): Hier ist scipy.optimize.minimize (BFGS/L-BFGS-B) dem Adam-Optimizer weit überlegen. BFGS nutzt die Krümmung der Kostenfunktion (Hesse-Matrix-Approximation) und konvergiert in wenigen Schritten präzise ("quadratische Konvergenz"). Adam (Gradient Descent) braucht tausende Schritte und "eiert" oft um das Optimum herum.
Online-Identifikation (Updates): Hier ist Adam (oder Recursive Least Squares - RLS) König, weil man nicht jedes Mal neu optimieren kann.
Was wirklich in math.py gehört (Der "Turbo" für OE)
Wenn du OE (Output Error) mit scipy.optimize nutzt, verbringt der Solver 90% der Zeit damit, den Gradienten zu schätzen (Finite Differenzen). Das bedeutet: Für jeden Schritt muss er N_{param} mal simulieren.
Du kannst das massiv beschleunigen (Faktor 20-50), indem du dem Solver den analytischen Gradienten gibst. Dafür brauchen wir Numba, weil wir rekursiv filtern müssen (Sensitivity Filtering).
Hier sind die zwei Snippets, die deine math.py braucht, um OE professionell schnell zu machen.
1. Der schnelle Simulator (Ersatz für lfilter)
Numba ist hier schneller als SciPy, weil wir den Overhead sparen und es direkt "inline" haben.
# math.py
import numpy as np
from numba import njit

@njit(cache=True)
def oe_simulate(u: np.ndarray, b: np.ndarray, f: np.ndarray, nk: int) -> np.ndarray:
    """
    Simuliert OE-Modell: y[k] = (B/F) * u[k-nk]
    y[k] = -f1*y[k-1]... + b0*u[k-nk]...
    """
    N = len(u)
    nf = len(f)
    nb = len(b)
    y = np.zeros(N)
    
    # f[0] ist immer 1.0 (wird ignoriert oder erzwungen)
    
    start = max(nf, nb + nk)
    
    for k in range(start, N):
        # B-Teil (Input)
        val = 0.0
        for j in range(nb):
            idx_u = k - nk - j
            if idx_u >= 0:
                val += b[j] * u[idx_u]
                
        # F-Teil (Feedback des Outputs)
        for i in range(1, nf):
            val -= f[i] * y[k - i]
            
        y[k] = val
        
    return y

2. Der "Alles-in-einem" Gradienten-Kernel (Das Herzstück)
Das hier ist Gold wert. Diese Funktion berechnet den Fehler (SSE) UND den Gradienten in einem Durchlauf. SciPy liebt das (jac=True).
Das Prinzip:
 * \frac{\partial \hat{y}}{\partial b_j} ist der Input u, gefiltert durch 1/F(q).
 * \frac{\partial \hat{y}}{\partial f_i} ist der negative Output -\hat{y}, gefiltert durch 1/F(q).

# math.py

@njit(cache=True)
def oe_cost_and_gradient(theta, u, y_true, nb, nf, nk):
    """
    Berechnet SSE (Sum Squared Error) und Gradientenvektor analytisch.
    theta Struktur: [b0, b1... | f1, f2...]
    """
    N = len(u)
    n_params = len(theta)
    
    # Parameter entpacken
    b = theta[:nb]
    f_coeffs = theta[nb:]
    # F-Polynom rekonstruieren: [1.0, f1, f2...]
    f = np.concatenate((np.array([1.0]), f_coeffs))
    
    # Arrays initialisieren
    y_sim = np.zeros(N)
    e = np.zeros(N)
    grad = np.zeros(n_params)
    
    # Sensitivity States (gefilterte Signale)
    u_filt = np.zeros(N) # u gefiltert mit 1/F
    y_filt = np.zeros(N) # y_sim gefiltert mit 1/F
    
    sse = 0.0
    
    start = max(nf, nb + nk)
    
    for k in range(start, N):
        # --- 1. Simulation step ---
        # Numerator (B)
        val_num = 0.0
        for j in range(nb):
            if k - nk - j >= 0:
                val_num += b[j] * u[k - nk - j]
        
        # Denominator (F) - Feedback
        val_den = 0.0
        for i in range(1, nf):
            val_den += f[i] * y_sim[k - i]
            
        y_sim[k] = val_num - val_den
        
        # Error
        err = y_true[k] - y_sim[k]
        sse += err * err
        
        # --- 2. Sensitivity Filtering (1/F) ---
        # Wir filtern u[k-nk] und y_sim[k] mit dem Pol-Polynom F
        
        # u_filt[k] = u[k-nk] - f1*u_filt[k-1]...
        curr_u_delayed = u[k-nk] if k-nk >= 0 else 0.0
        val_uf = curr_u_delayed
        for i in range(1, nf):
            val_uf -= f[i] * u_filt[k-i]
        u_filt[k] = val_uf
        
        # y_filt[k] = y_sim[k] - f1*y_filt[k-1]...
        # WICHTIG: Hier nutzen wir y_sim als Input für den Sens-Filter
        val_yf = y_sim[k] 
        for i in range(1, nf):
            val_yf -= f[i] * y_filt[k-i]
        y_filt[k] = val_yf
        
        # --- 3. Gradient Accumulation ---
        # dJ/dtheta = -2 * e[k] * dy/dtheta
        
        # Gradient b_j: dy/db_j = u_filt[k-j] (weil B ein FIR Filter auf dem gefilterten u ist)
        for j in range(nb):
            if k - j >= 0:
                sens = u_filt[k - j] 
                grad[j] += -2 * err * sens
                
        # Gradient f_i: dy/df_i = -y_filt[k-i]
        for i in range(1, nf):
            idx_grad = nb + (i - 1)
            sens = -y_filt[k - i]
            grad[idx_grad] += -2 * err * sens
            
    return sse, grad

Integration in deine Klasse OE
Das macht deine OE-Klasse extrem sauber und schnell. Du übergibst die Numba-Funktion direkt an den Minimierer.
class OE(PEM):
    # ... init ...

    def _ident(self, order: Tuple[int, int, int]) -> LTIModel:
        nb, nf, nk = order
        
        # Initialisierung (z.B. via ARX oder einfache Annahme)
        # mod = ...
        
        # Startwerte extrahieren [b..., f...]
        # Achtung: f startet bei f1, da f0=1 fest ist
        theta0 = np.concatenate((mod.b, mod.a[1:])) 
        
        # Optimierung mit analytischem Gradienten
        def objective(theta):
            return _math.oe_cost_and_gradient(
                theta, self.u, self.y, nb, nf + 1, nk # nf+1 weil math.py len(f) erwartet (inkl 1.0)
            )
            
        res = scipy.optimize.minimize(
            objective, 
            theta0, 
            method='BFGS', 
            jac=True, # <--- DAS ist der Turbo!
            options={'disp': True}
        )
        
        # Ergebnis zurückschreiben
        theta_opt = res.x
        mod.b = theta_opt[:nb]
        mod.a = np.concatenate(([1.0], theta_opt[nb:]))
        
        return mod

Ergebnis:
Statt 20-50 Simulationen pro Schritt (Finite Differenzen) machst du jetzt nur noch 1 Simulation + Filterung. Das ist mathematisch exakt und programmiertechnisch hocheffizient.
Das ist das einzige Numba-Snippet, das du wirklich brauchst.
